### ISSUE TITLE
Use python type in Mapped to boolean expression catching thinks like `Mapped[int] == 'a string'`

### ISSUE THREAD

**Issue by ledz911** [sqlalchemy/sqlalchemy#9587]
### Ensure stubs packages are not installed

- [X] No sqlalchemy stub packages is installed (both `sqlalchemy-stubs` and `sqlalchemy2-stubs` are not compatible with v2)

### Verify if the api is typed

- [X] The api is not in a module listed in [#6810](https://github.com/sqlalchemy/sqlalchemy/issues/6810) so it should pass type checking

### Confirm this is not Mypy dataclasses issue 13856

- [X] This is not related to Mypy issue 13856

### Describe the typing issue

Hello,
why when I execute a query pyright does not work? I'm trying to compare and integer with a string. 

`stmt = select(User).where(User.age == 'test')      # why doesn't pyright fail?`

Thanks.

### To Reproduce

```python
from sqlalchemy import select, create_engine
from sqlalchemy.orm import Session
from sqlalchemy import String
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column

engine = create_engine("sqlite://", echo=True)
session = Session(engine)


class Base(DeclarativeBase):
    pass


class User(Base):
    __tablename__ = "users"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(30))
    age: Mapped[int]


stmt = select(User).where(User.age == 'test')      # why doesn't pyright fail?
```

### Versions

- OS: MacOS 12.6.3
- Python: 3.11.1
- SQLAlchemy: 2.0.8
- Type checker: Pylance v2023.1.10

**Comment by CaselIT** [comment]
Hi,

This is not something that is yet supported by typing, and I'm not sure if it's something that can be generally supported given the current python typing capabilities.

Also there may be cases where you may want to use a different python type than the one specified in the column.

Still it would be something nice to have

**Comment by CaselIT** [comment]
In principle this may actually be doable.
The main thing is coming up with a type to place instead of the `other: Any` in `SQLCoreOperations`

https://github.com/sqlalchemy/sqlalchemy/blob/15e903d77df78ed2a6e63c05ce7c26f71a4c33c4/lib/sqlalchemy/sql/elements.py#L796-L812

I have not verified if there are other limitation in the type checker though

**Comment by ledz911** [comment]
First of all thank you so much for replying @CaselIT!

I replaced this piece of code:
```
def __lt__(self, other: Any) -> BooleanClauseList: 
     ... 
```

With this piece of code:
```
def __lt__(self, other: _T) -> BooleanClauseList: 
     ... 
```

I also changed _in operator:
```
def in_(
            self,
            other: Union[
                Iterable[_T], BindParameter[Any], roles.InElementRole
            ],
        ) -> BinaryExpression[bool]:
            ...
```

And now pyright fails!

Of course I have no idea if this change involves global changes to sqlalchemy, so I ask you if it would be feasible.
Thanks

**Comment by CaselIT** [comment]
That's good to know that type checker can support this.

Using `_T` is not enough, since you would get an error if you were to write something like `User.is == User.age` (that does not make much sense, just an example of comparison of sqlalchemy types) but that's a start.
Off the top of my head I don't know what the most appropriate type would be to support the above case, but I believe we may already have an alias for this, we just need to look

**Comment by ledz911** [comment]
@CaselIT actually if I write:
```
User.id == User.age      # no error: OK
User.id == User.name   # no error: FAIL -> I'm trying to compare int with str
```

**Comment by CaselIT** [comment]
yes, that's the general idea, but I'm not sure if `User.id == User.name` should in general error.
there may be cases where it may make sense to allow such an operation (maybe not for `==` but for other operators)

**Comment by zzzeek** [comment]
apparently there's a class of error in mypy called "comparison-overlap", which reports for this:

```py
x: str = "a"
y: int = 2

print(x == y)
```

however, the above program does not report a runtime error, it is perfectly legal to use ``==`` for any two types, that's part of the contract.   similarly in SQL it's not consistently invalid to compare different types without casts

mariadb:

```
MariaDB [test]> select 1 = 'hi';
+----------+
| 1 = 'hi' |
+----------+
|        0 |
+----------+
1 row in set, 1 warning (0.002 sec)

```

sqlite:

```
sqlite> select 1 = 'hi';
0

```

Postgresql of course decides it doesn't like this:

```
test=# select 1 = 'hi';
ERROR:  invalid input syntax for type integer: "hi"

```


I think within SQLAlchemy using _T here might have produced a lot of errors in the library's typing since the "==" thing works fine for two ColumnElements that have different types and we also use them in hash/mapping contexts and such where we rely upon comparison operators to not care about the internal type.

since mypy does have an error here I'm not opposed to trying to change it again but it might not be feasible.

**Comment by CaselIT** [comment]
I can see something like this as being something some people may want and some people may not want.

I don't know if there is the ability of turning on/off this behaviour, maybe offering a something like a subclass of `Mapped` that has stricter typing cold be an option? Something like `StrictMapped` and `strict_mapped_column` (or `mapped_column(..., strict=True)`)

From a maintenance point of view it should not be that large, but of course it's yet another thing user need to reason about, so not sure if it's worth it


Also in some cases it's hard to type correctly, like `>` even when strictly types should allow `something_int > something_float`, and I don't know if we can to that in a generic way without hard-coding a lot of overloads

**Comment by zzzeek** [comment]
this behavior is much lower level than Mapped and it would not be my own preference to just have an override at the Mapped level, I'd want to be able to swap out all of ColumnOperators since that's where this happens.

Mapped is though where it will usually be most apparent, but as mentioned it's really not invalid to compare different types in SQL in most cases, and making this more strict can make typing more difficult.      Like if I wanted to compare `Mapped[int] > func.foobar()`, and usually i dont have to be bothered adding `type_=Integer` to `func.foobar()`, now I do.  Unless we got the `__eq__()` to also accommodate ``NullType`` / ``Optional[_T]``.    I think there are a lot of different cases here with the comparison operators.

**Comment by CaselIT** [comment]
> this behavior is much lower level than Mapped and it would not be my own preference to just have an override at the Mapped level, I'd want to be able to swap out all of ColumnOperators since that's where this happens.
> Mapped is though where it will usually be most apparent,

for sure mapped is not where that happen, but it's the most prominent place where people would want/expect this behaviour.
Regardless of how it's implemented doing it more generally would mean having a lot of duplicate classes, unless we find a typing switch that could be used for this (I'm not aware of any though)

Overall I agree with your other observation, doing this without being too strict may not be easy.

**Comment by zzzeek** [comment]
at the moment it feels ambitious like a 2.1 thing if we were motivated.  we want to get the current typing platform very solid before tightening more screws.

**Comment by CaselIT** [comment]
I think before thinking at how it would be implemented we would need a plan on what to do.

If there is no other option than using a different subclass, I don't think it's reasonable to have "StrictThing" for each "Thing" that has column operator behaviour.
